### ISSUE TITLE
[RFC] Terrains

### ISSUE THREAD

**Issue by jojolepro** [amethyst/amethyst#653]
Should terrains be made in an external tool?
Should it be limited to a single plane per xz position?

How do we manage texture merging between different heights?

This description will be modified after discussion.

**Comment by OvermindDL1** [comment]
Don't forget, not just hightmap terrains but also voxels of many forms, different splatting styles, chunk based or not, generative (infinite) or not, LODs (important to consider with generative LODs) or not, skirts or not, etc etc etc...  Giant massive topic here..

**Comment by Xaeroxe** [comment]
I'd say voxels are out of scope for this RFC, but also definitely a goal of Amethyst outside this RFC.  Procedurally generated terrain should be built on top of whatever we make here, but for now I'd also call it out of scope.

**Comment by torkleyy** [comment]
> Should terrains be made in an external tool?

We should use a (if possible: common) file format. That's all. You can use
an external tool, maybe we'll create our own tool one day, but if you got
too much time there should be no restriction not to write it by hand. The
point is it should be an open and easy to understand format, which tool you
use doesn't matter.

But before we can talk about file formats, we need to actually support the
rendering of terrain ;)

**Comment by Xaeroxe** [comment]
>But before we can talk about file formats, we need to actually support the
rendering of terrain ;)

I think we basically already do?  It's just a mesh with a texture on it right?

**Comment by AnneKitsune** [comment]
For a simple one, its just a grid of triangles with a generated texture

**Comment by torkleyy** [comment]
There's a lot more to terrain rendering:

* heightmaps
* materials
* water
* grass
* fog, clouds
* instanced objects (trees, bushes, etc)

**Comment by Xaeroxe** [comment]
Yeah that's true.  Before we can render anything we at least need some `struct`s to stick the data in though

**Comment by AnneKitsune** [comment]
Water->Not actually related to terrain, most game implements this by adding water "boxes" around and through the terrain's ground.
Grass->Considered "foliage" (#654 )
Instanced objects-> Foliage (#654 )
Fog + Clouds->Separated from terrain.
Heightmaps->What I called Texture merging

Updating name to be less confusing as to what "terrain" is.

**Comment by OvermindDL1** [comment]
So should this be renamed to hightmaps then as it's not generic terrain?  Don't forget that hightmaps have non terrain usefulness as well.

As for just hightmap rendering, I've generally always had a texture array for the splats, a 16-bit grayscale image (or 32bit but I've never needed that much space) for the height, an rgba image where each channel represents what splat mixture to use at that point, a normal image for fast pathfinding (not lighting, this normal image is for faster pathfinding and it may not match the actual angles of the map), and occasionally another rgba image for indexing into the splat array (so this one would define what splat texture to use at this location and the splat mixture image defines how mix the 4 selected images at this point this meaning you can have up to 256 images (atlas I used) on a single heightmap with up to 4 mixing at any specific point, trivial to do in shader).  I thought I had another as well but can't recall off hand...

Then of course a Heightmap-based terrain can be built using the above primitive, but definitely keep the above primitive useful standalone as they have uses other than for generic heightmap terrains (especially once transformed/rotated).

**Comment by OvermindDL1** [comment]
There could of course be more modern rendering styles, the above is just how I made mine many many years ago all in textures with a single flat mesh passed in.

**Comment by AnneKitsune** [comment]
Well, since it can have multiple heights per xz position (a cave, for example), we can't really call it a heightmap?

**Comment by OvermindDL1** [comment]
How are those planned to work, multiple layers of heightmaps alternating flipped to allow for multiple internal layers so things like caves are done, or planning for sections to become invisible and replaced with custom meshes of any format entirely (I've done both in the past), or something else?

**Comment by AnneKitsune** [comment]
I actually didn't think much about how caves would be done. I know that not having them is a massive limitation of most engines. It requires devs to makes holes in the terrain, and then manually add custom meshes.

**Comment by OvermindDL1** [comment]
Custom meshes aren't bad though, they can look fantastic if done right.  Layered heightmaps work fine but look downright funky at times because the map itself can't angle properly without starting to add tons and tons of layers (aren't you making a voxel thing at that point anyway?).  If you want proper caves and you don't want custom meshes then really the only good option then are voxels (voxels can do an excellent job replacing heightmaps, can even just not render the bottom-most polys as well, honestly I'd not mess with heightmaps at all if I were to remake my engine today because you can still load a heightmap into a voxel system anyway so I'd focus on making the best voxel system properly, and with a simple RLE optimization they aren't really any more costly than heightmaps anyway in the case of heightmaps).

**Comment by torkleyy** [comment]
I agree with @OvermindDL1.

Btw, don't forget lighting, that's also special for terrains.

**Comment by AnneKitsune** [comment]
So we agree to remove the cave feature, and have holes instead?
Or is there a better way than to have holes directly into the mesh?

**Comment by OvermindDL1** [comment]
Holes could easily be done by a single bit per image or via an invisible splat layer.  I've only ever used custom meshes on holes or layered heightmaps, I don't really know of any other way of hand.  Still though, I'd vote for just going pure voxels anyway as they can represent heightmaps too, you can always add special detection to render slightly more efficiently in the case of rendered chunks being heightmap-like, that still allows other rendered chunks to fall back to proper full voxels for caves or whatever anyway.  In addition this would allow for special rendering as well like a game could have normal angled terrain, or could use square blocks, or hexes, or whatever visualizer they want for the voxel rendering (I always did multiple overlapping voxel layers specialized for different shapes).

**Comment by OvermindDL1** [comment]
Otherwise I'd just say support holes, they support whatever is needed regardless anyway.

**Comment by AnneKitsune** [comment]
Okay, I did some research on voxel engines and that approach looks promising. My only concern with it is that they actually look a bit blocky, even at high resolutions. This one for example: https://www.youtube.com/watch?v=4oGl3F8xf6M or this https://www.youtube.com/watch?v=lXm5JWys55o

Also, voxel worlds appear to be a lot more worse performance wise, from what I saw.

If there's a way to make voxel worlds look realistic and decently performant, then I'm okay with going for a voxel only approach.

I think that it would be possible to use voxels for only the terrain (read ground), use instanced objects made in an external tool (trees) and also have foliage (3d grass or 2d billboard with a picture of grass). For fluids, I found something about "fluid dynamics" which allows voxels to act as fluid. If this consumes too much power, or doesn't look good, there's always the possibility of using other solutions like mesh deformations for waves. 

All this would allow the user to either:
- Go full voxel
- Use voxel for the terrain, then use 3d objects for buildings, trees etc
- Use only 3d objects (but then we don't have a 3d ground editor)

So, is this a viable option?

**Comment by Xaeroxe** [comment]
Voxel is great for some games but it is far from a catch-all solution.  I'd be strongly against supporting voxel only, however I do eventually want a voxel option.

**Comment by AnneKitsune** [comment]
https://www.youtube.com/watch?v=XkSS_veoSg0

**Comment by AnneKitsune** [comment]
Would multiple variable length voxel do the trick for a heightmap equivalent?
You'd have the voxel chunk with a fixed xz size, but the y size would be dynamic depending on the minimum and maximum heights of the terrain inside that one chunk.
If at all possible, it would be nice to have reusable solution for both voxel blocky worlds as well as super smooth terrains with caves. Also, being able to break any part of the world with explosion is pretty neat too :P

**Comment by Xaeroxe** [comment]
Alright my position has definitely changed after seeing that tech demo.  I still think voxel isn't the most efficient way to do that but it certainly can be good enough for now.

**Comment by OvermindDL1** [comment]
Voxel is blocky depending on the renderer.  It can render as blocks, marching cubes, smooth splines, all sorts of things.  And they are not worse in performance, they are actually quite efficient...  And yes, things like trees and such are generally not voxels, just terrain and so forth (though some games do have trees be voxels too, but eh).

**Comment by AnneKitsune** [comment]
Thoughts? @omni-viral @torkleyy @Rhuagh

**Comment by MrMinimal** [comment]
Being used to Quad trees for planet rendering, I am really impressed by [how far commercial voxel engines](https://www.youtube.com/watch?v=nr5JqYYye3w) have come.

**Comment by AnneKitsune** [comment]
That's actually insane.

**Comment by erlend-sh** [comment]
my 2c:

> Should terrains be made in an external tool?

Yes. A general purpose terrain system is far from an essential utility in a game engine's repertoire. It's great for quick prototypes (lowers the time-to-interesting-interaction) and demonstrating the graphics ability of the engine. Even so, most games will end up building their own tooling for controlling the entire play-space as a whole. The way you construct the game space tends to be such an integral aspect of the creative process that most game designers strongly prefer to roll their own tooling for "game space creation".

There's also a lot of different ways to go about making a terrain system, and whichever you choose will always be pulling the rest of the engine a little bit in that direction. I think the best approach to terrain systems is to leave it completely to plugin territory, keeping the rendering stack as unopinionated as possible.

**Comment by fhaynes** [comment]
Moving to RFC repo
